### PR TITLE
[ceres + downstreams] Update to ceres to 2.2.0

### PR DESCRIPTION
**Pull request by ahojnnes**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by ahojnnes** [comment]
@microsoft-github-policy-service agree [company="{your company}"]

**Comment by ahojnnes** [comment]
Build failures appear to be unrelated to this change? `vtk[core]:x64-windows` seems to not exist?

**Comment by BillyONeal** [comment]
> Build failures appear to be unrelated to this change? `vtk[core]:x64-windows` seems to not exist?

They look legitimate to me? A couple of examples.

Direct failures:

```
REGRESSION: ceres:x86-windows failed with BUILD_FAILED. If expected, add ceres:x86-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
REGRESSION: ceres:x64-windows failed with BUILD_FAILED. If expected, add ceres:x64-windows=fail to D:\a\_work\1\s\scripts\azure-pipelines/../ci.baseline.txt.
```

Indirect failures, for instance:

```
REGRESSION: cartographer:arm64-osx failed with BUILD_FAILED. If expected, add cartographer:arm64-osx=fail to /Users/vcpkg/Data/work/1/s/scripts/azure-pipelines/../ci.baseline.txt.
REGRESSION: openturns:arm64-osx failed with BUILD_FAILED. If expected, add openturns:arm64-osx=fail to /Users/vcpkg/Data/work/1/s/scripts/azure-pipelines/../ci.baseline.txt.
REGRESSION: openmvg:arm64-osx failed with BUILD_FAILED. If expected, add openmvg:arm64-osx=fail to /Users/vcpkg/Data/work/1/s/scripts/azure-pipelines/../ci.baseline.txt.
REGRESSION: vcpkg-ci-itk:arm64-osx cascaded, but it is required to pass. (/Users/vcpkg/Data/work/1/s/scripts/azure-pipelines/../ci.baseline.txt).
REGRESSION: vcpkg-ci-paraview:arm64-osx cascaded, but it is required to pass. (/Users/vcpkg/Data/work/1/s/scripts/azure-pipelines/../ci.baseline.txt).
```

are downstream broken components:

```console
PS D:\vcpkg> .\vcpkg.exe depend-info cartographer 2>&1 | findstr ceres
ceres[lapack, suitesparse]: eigen3, glog, lapack, suitesparse, vcpkg-cmake, vcpkg-cmake-config
cartographer: boost-iostreams, cairo, ceres, gflags, glog, gtest, lua, protobuf, vcpkg-cmake, vcpkg-cmake-config
PS D:\vcpkg> .\vcpkg.exe depend-info openturns 2>&1 | findstr ceres   
ceres: eigen3, glog, vcpkg-cmake, vcpkg-cmake-config
openturns: blas, boost-multiprecision, boost-random, ceres, cminpack, dlib, exprtk, hdf5, lapack, libxml2, mpc, mpfr, muparser, nlopt, pagmo2, pthread, python3, spectra, tbb, vcpkg-cmake, vcpkg-cmake-config
PS D:\vcpkg> .\vcpkg.exe depend-info openmvg 2>&1 | findstr ceres     
ceres[cxsparse, lapack, suitesparse]: eigen3, glog, lapack, suitesparse, vcpkg-cmake, vcpkg-cmake-config
openmvg: cereal, ceres, coin-or-clp, coin-or-osi, coinutils, eigen3, flann, libjpeg-turbo, liblemon, libpng, tiff, vcpkg-cmake, vcpkg-cmake-config, vlfeat, zlib
PS D:\vcpkg> 
```

File: versions/o-/openmvg.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "e08aec4337f9a281817b2c84d04fee5a079551d8",
+      "git-tree": "ce150540ea4d0c70dbef977eac1c630c532f0ec9",
```
Comment (JonLiu1993) [inline versions/o-/openmvg.json:4]: @ahojnnes, Thanks for your PR, please revert the history git-tree, don't modify it, just add the new one.

Comment (ahojnnes) [reply versions/o-/openmvg.json]: I hope it's correct now. Thanks.

**Comment by patrikhuber** [comment]
Thank you Johannes, that is a huge amount of work, especially having to fix all of upstream too that still depends on old Ceres syntax.

Could I just ask, could you confirm that we can still use Ceres with Eigensparse but without METIS, as I've described here 
 https://github.com/microsoft/vcpkg/pull/35021#issuecomment-1851871588?

Thank you!

**Comment by ahojnnes** [comment]
@patrikhuber If you don't select the "ceres[suitesparse]" feature, ceres can use the eigen sparse solver without the need for metis.

**Comment by ahojnnes** [comment]
@BillyONeal @JonLiu1993 This has been a deeper rabbit hole than expected. This should be ready to review and merge.

**Comment by patrikhuber** [comment]
> @patrikhuber If you don't select the "ceres[suitesparse]" feature, ceres can use the eigen sparse solver without the need for metis.

Thank you, that's fantastic!

**Comment by ahojnnes** [comment]
@pmloulon FYI. I did a best effort in updating to latest openmvg. You may want to apply some of the patches upstream. Most of them should be useful independent of vcpkg.

File: versions/o-/openmvg.json [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "28c8872842fd1bac86010c5a235831d25b6f8add",
+      "version": "2.1",
+      "port-version": 0
+    },
+    {
+      "git-tree": "4d295167110da197b9fa74056de4c0863d1953c9",
+      "version": "2.0",
+      "port-version": 12
+    },
```
Comment (JonLiu1993) [inline versions/o-/openmvg.json:12]:
```suggestion
```

**Comment by BillyONeal** [comment]
Hi there!

Thanks for all the downstream fixes! Unfortunately the resulting patches are huge :(. Can you make sure these are submitted upstream? With changes this big we want to give them an opportunity to respond before 'putting words in their mouth'. Thanks!

File: versions/r-/rtabmap.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "1161d6eb7dacc4aa2cff53381126c89c87f30fe0",
+      "git-tree": "1724fe107ebd054a133ff197201900a136d60d0b",
```
Comment (JonLiu1993) [inline versions/r-/rtabmap.json:4]: Pleasae don't change the history "git-tree"

File: versions/t-/theia.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "versions": [
     {
-      "git-tree": "fcce9342a3c7301139dd9d9a388657d99fdcc159",
+      "git-tree": "eb1ed0cff079b1deddd5dccde7c5199d4dd01618",
```
Comment (JonLiu1993) [inline versions/t-/theia.json:4]: Same as above

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Comment by ahojnnes** [comment]
> Hi there!
> 
> Thanks for all the downstream fixes! Unfortunately the resulting patches are huge :(. Can you make sure these are submitted upstream? With changes this big we want to give them an opportunity to respond before 'putting words in their mouth'. Thanks!

The patches look bigger than they are. They are mostly the same but with the following changes: (1) Some of the patches were combined into a single logical patch instead of applying a chain of patches on top of the same files (which becomes very hard to maintain). (2) Other changes are incurred by slightly changed underlying source code as the updated library versions had some small changes in the files themselves. The reason I went with updating the downstream library versions was that some of the breaking changes by Ceres have already been addressed by the upstream repos. Otherwise, the patches in this PR would have had much bigger changes.

As for upstreaming the patches, the plan was to follow up for openmvg (whose maintainer @pmoulon I already tagged above), rtabmap and openturns. However, theia and cartographer are both unmaintained codebases. I don't think anybody will merge these. I would argue that these shouldn't be packages in vcpkg in the first place, given the maintenance burden they impose on the contributors here.

File: versions/t-/theia.json (hunk 4 above)
Comment (ahojnnes) [reply versions/t-/theia.json]: Thank you, done.

**Review by JonLiu1993 — APPROVED** [review]

**Comment by dg0yt** [comment]
> CMake Error at F:/Feature-test/ceres/scripts/buildsystems/vcpkg.cmake:639 (_add_library):
>   _add_library cannot create ALIAS target "gflags" because another target
>   with the same name already exists.
> Call Stack (most recent call first):
>   CMakeLists.txt:402 (add_library)

This look like an error in your test project.

**Comment by ahojnnes** [comment]
@JonLiu1993 This should be fixed now.

**Comment by ahojnnes** [comment]
@BillyONeal I am also working on unrolling the patches into a series of upstream pull requests, but this will take time and largely also depends on the maintainers.

**Comment by JonLiu1993** [comment]
Whene I tested the usage of `cartographer`, I get thie error, please take a look:
CMakeLists.txt:
````
cmake_minimum_required(VERSION 3.5)

project(test)

find_package(cartographer CONFIG REQUIRED)

add_executable(test CMakeProject1.cpp)

target_link_libraries(test PRIVATE cartographer)
````
Cofig:
````
1> CMake generation started for default configuration: 'x64-Debug'.
1> Found and using vcpkg toolchain file (F:/Feature-test/ceres/scripts/buildsystems/vcpkg.cmake).
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/ceres/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"
1> Working directory: C:\Users\test\source\repos\CMakeProject1\out\build\x64-Debug
1> [CMake] -- The C compiler identification is MSVC 19.41.34123.0
1> [CMake] -- The CXX compiler identification is MSVC 19.41.34123.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] -- Looking for sgemm_
1> [CMake] -- Looking for sgemm_ - found
1> [CMake] -- Found BLAS: F:/Feature-test/ceres/installed/x64-windows/debug/lib/openblas.lib
1> [CMake] -- Using VCPKG FindLAPACK from package 'lapack-reference'
1> [CMake] -- Looking for cheev_
1> [CMake] -- Looking for cheev_ - found
1> [CMake] -- A library with LAPACK API found.
1> [CMake] -- Found CUDAToolkit: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v12.5/include (found suitable version "12.5.40", minimum required is "12.5.40")
1> [CMake] -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
1> [CMake] -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
1> [CMake] -- Looking for pthread_create in pthreads
1> [CMake] -- Looking for pthread_create in pthreads - not found
1> [CMake] -- Looking for pthread_create in pthread
1> [CMake] -- Looking for pthread_create in pthread - not found
1> [CMake] -- Found Threads: TRUE
1> [CMake] -- Found required Ceres dependency: Eigen version 3.4.0 in F:/Feature-test/ceres/installed/x64-windows/share/eigen3
1> [CMake] -- Found Ceres version: 2.2.0 installed in: F:/Feature-test/ceres/installed with components: [EigenSparse, SparseLinearAlgebraLibrary, LAPACK, SuiteSparse, SchurSpecializations]
1> [CMake] -- Configuring done (15.8s)
1> [CMake] CMake Error at F:/Feature-test/ceres/installed/x64-windows/share/cartographer/CartographerTargets.cmake:61 (set_target_properties):
1> [CMake]   The link interface of target "cartographer" contains:
1> [CMake] 
1> [CMake]     Boost::iostreams
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
1> [CMake] 
1> [CMake] Call Stack (most recent call first):
1> [CMake]   F:/Feature-test/ceres/installed/x64-windows/share/cartographer/cartographer-config.cmake:72 (include)
1> [CMake]   F:/Feature-test/ceres/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   CMakeProject1/CMakeLists.txt:7 (find_package)
1> [CMake] -- Generating done (0.0s)
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/ceres/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/ceres/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"' returned with exit code: 1'.

````

**Comment by ahojnnes** [comment]
@JonLiu1993 Thanks for finding the issue. This should also be fixed now.

**Comment by BillyONeal** [comment]
> The patches look bigger than they are. They are mostly the same but with the following changes: (1) Some of the patches were combined into a single logical patch instead of applying a chain of patches on top of the same files (which becomes very hard to maintain). 

What matters is the total size of the patches, not how many files it's split into; I agree separating them would be bad.

> However, theia and cartographer are both unmaintained codebases.

Hmmmm.... they also have very high port-versions suggesting that they are painful. Perhaps we should deindex them.

**Comment by ahojnnes** [comment]
> > The patches look bigger than they are. They are mostly the same but with the following changes: (1) Some of the patches were combined into a single logical patch instead of applying a chain of patches on top of the same files (which becomes very hard to maintain). 
> 
> 
> 
> What matters is the total size of the patches, not how many files it's split into; I agree separating them would be bad.

That's not what I meant to say. To clarify: what I meant is that I only added a relatively small delta of new changes to the existing patches. The main change I made was to reapply them to the updated underlying sources and combined them into fewer patch files. The only project with somewhat significant changes is rtabmap, for which I already upstreamed the changes in a PR. It will take time for these to be consumable here though. I would argue that, if we were fine accepting these patches in the past, then we should be fine again now. It would be unfortunate if we delay the update of ceres much longer. Many external projects rely on it and would greatly benefit from this update.

I still recommend decommissioning the theia and cartographer ports, given that the upstream projects are not maintained anymore.

**Comment by BillyONeal** [comment]
>That's not what I meant to say. To clarify: what I meant is that I only added a relatively small delta of new changes to the existing patches.

I guess, what I mean is, I don't think we should have accepted a 1500 line patch in the first place. That's well outside of our normal patch guidelines, and it's unwarranted if there is no universe where upstream could someday accept it.

>I still recommend decommissioning the theia and cartographer ports, given that the upstream projects are not maintained anymore.

Yes, those are the two in question here. @AugP @ras0219-msft @JavierMatosD @vicroms and I discussed today and agree these should be removed from the index. I'll push that into this PR shortly.

**Comment by BillyONeal** [comment]
> Whene I tested the usage of `cartographer`, I get thie error, please take a look:

I deindexed cartographer entirely and have an action item to update the maintainer guide about these cases where large patches are necessary to get unmaintained downstream dependents functional. Do you have other concerns here @JonLiu1993 ?

@ahojnnes Are you happy with this 'the ports got deindexed' outcome?

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
Tested usage successfully by `ceres:x64-windows`
All features are tested successfully in the following triplet:

- x86-windows
- x64-windows

File: ports/openmvg/portfile.cmake [hunk 5]
```diff
@@ -77,17 +75,22 @@ vcpkg_cmake_configure(
...
-vcpkg_cmake_config_fixup(CONFIG_PATH share/openMVG/cmake)
 
-if (NOT VCPKG_BUILD_TYPE OR VCPKG_BUILD_TYPE STREQUAL "debug")
-    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" "${CURRENT_PACKAGES_DIR}/debug/share")
-    file(REMOVE "${CURRENT_PACKAGES_DIR}/debug/lib/openMVG-targets.cmake" "${CURRENT_PACKAGES_DIR}/debug/lib/openMVG-targets-debug.cmake")
-endif()
-file(REMOVE "${CURRENT_PACKAGES_DIR}/lib/openMVG-targets.cmake" "${CURRENT_PACKAGES_DIR}/lib/openMVG-targets-release.cmake")
+make_directory("${CURRENT_PACKAGES_DIR}/share/openMVG/")
+file(RENAME "${CURRENT_PACKAGES_DIR}/lib/openMVG/cmake" "${CURRENT_PACKAGES_DIR}/share/openMVG/cmake")
+make_directory("${CURRENT_PACKAGES_DIR}/debug/share/openMVG/")
```
Comment (dg0yt) [inline ports/openmvg/portfile.cmake:81]:
```suggestion
file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/share/openMVG/")
file(RENAME "${CURRENT_PACKAGES_DIR}/lib/openMVG/cmake" "${CURRENT_PACKAGES_DIR}/share/openMVG/cmake")
file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/debug/share/openMVG/")
```

`make_directory` is deprecated.

**Comment by ahojnnes** [comment]
> > Whene I tested the usage of `cartographer`, I get thie error, please take a look:
> 
> I deindexed cartographer entirely and have an action item to update the maintainer guide about these cases where large patches are necessary to get unmaintained downstream dependents functional. Do you have other concerns here @JonLiu1993 ?
> 
> @ahojnnes Are you happy with this 'the ports got deindexed' outcome?

Thank you, sounds good to me!

**Comment by JonLiu1993** [comment]
Looks there is a error in x64-linux, this is error log:
[install-x64-linux-dbg-out.log](https://github.com/user-attachments/files/18093540/install-x64-linux-dbg-out.log)

**Comment by ahojnnes** [comment]
@JonLiu1993 Shouldn't have removed the openmvg:x64-linux from the ci exclusions :-) I'll see if I can fix it, but otherwise will revert back to the previous state and skip it.

**Comment by ahojnnes** [comment]
@JonLiu1993 @BillyONeal All checks should be green now. Thank you for your reviews and support.

**Review by JonLiu1993 — APPROVED** [review]

**Review by JavierMatosD — COMMENTED** [review]
LGTM, please fix the file conflicts

**Comment by ahojnnes** [comment]
@JavierMatosD @BillyONeal @JonLiu1993 I'd appreciate if we could merge this update, so I don't run into further merge conflicts. Thank you.

**Comment by patrikhuber** [comment]
That is amazing, thank you very much @ahojnnes for the work!

**Pull request merged**